File: C++/Medium/Accepted/filling-bookcase-shelves_8.cpp
class Solution {
public:
    int minHeightShelves(vector<vector<int>>& books, int shelfWidth) {
        
        int n = books.size();
        vector<int> dp(n+1, 0);
        dp[0] = 0;
        for (int i = 0; i < n; ++i) {
            int width = books[i][0];
            int height = books[i][1];
            dp[i+1] = dp[i] + height;
            for (int j = i-1; j >= 0; --j) {
                width += books[j][0];
                height = max(height, books[j][1]);
                if (width > shelfWidth)
                    break;
                dp[i+1] = min(dp[i+1], dp[j] + height);
            }
        }
        return dp[n];
    }
};

File: Python3/Medium/Accepted/filling-bookcase-shelves_8.py
class Solution:
    def minHeightShelves(self, books: List[List[int]], shelfWidth: int) -> int:
        
        # dp[i] = min(dp[i], dp[j] + max(height[j+1..i]))
        n = len(books)
        dp = [0] + [float('inf')] * n
        for i in range(n):
            width, height = 0, 0
            for j in range(i, -1, -1):
                width += books[j][0]
                if width > shelfWidth: break
                height = max(height, books[j][1])
                dp[i+1] = min(dp[i+1], dp[j] + height)
        return dp[-1]